crates/miden-objects/src/account/component/template/storage/toml.rs

```diff
@@ -135,146 +175,151 @@ impl<'de> serde::Deserialize<'de> for WordRepresentation {
-        match self {
-            StorageValues::Words(vec) => Some(vec),
-            StorageValues::MapEntries(_) => None,
-        }
-    }
-
-    pub fn into_map_entries(self) -> Option<MapRepresentation> {
-        match self {
-            StorageValues::Words(_) => None,
-            StorageValues::MapEntries(map) => Some(map),
-        }
-    }
-
-    pub fn len(&self) -> Option<usize> {
-        match self {
-            StorageValues::Words(vec) => Some(vec.len()),
-            StorageValues::MapEntries(map) => map.len(),
-        }
-    }
+    MapEntries(Vec<MapEntry>),
 }
 
 // STORAGE ENTRY SERIALIZATION
 // ================================================================================================
 
-/// Used as a helper for validating and (de)serializing storage entries
-#[derive(Default, Deserialize, Serialize)]
+#[derive(Default, Debug, Deserialize, Serialize)]
 struct RawStorageEntry {
-    name: String,
+    name: Option<String>,
     description: Option<String>,
     slot: Option<u8>,
     slots: Option<Vec<u8>>,
-    value: Option<WordRepresentation>,
+    #[serde(rename = "type")]
+    word_type: Option<String>,
+    value: Option<[FeltRepresentation; 4]>,
     values: Option<StorageValues>,
 }
 
 impl From<StorageEntry> for RawStorageEntry {
     fn from(entry: StorageEntry) -> Self {
         match entry {
-            StorageEntry::Value { name, description, slot, value } => RawStorageEntry {
-                name,
-                description,
-                slot: Some(slot),
-                value: Some(value),
-                ..Default::default()
+            StorageEntry::Value { slot, word_entry } => match word_entry {
+                WordRepresentation::Value { name, description, value } => RawStorageEntry {
+                    slot: Some(slot),
+                    name: name.as_ref().map(StorageValueName::to_string),
+                    description: description.map(String::from),
+                    value: Some(value),
+                    ..Default::default()
+                },
+                WordRepresentation::Template { name, description, r#type } => RawStorageEntry {
+                    slot: Some(slot),
+                    name: Some(name.to_string()),
+                    description: description.map(String::from),
+                    word_type: Some(r#type),
+                    ..Default::default()
+                },
             },
-            StorageEntry::Map { name, description, slot, map: values } => RawStorageEntry {
-                name,
-                description,
+            StorageEntry::Map { slot, map } => RawStorageEntry {
+                name: Some(map.name().to_string()),
+                description: map.description().cloned(),
                 slot: Some(slot),
-                values: Some(StorageValues::MapEntries(values)),
+                values: Some(StorageValues::MapEntries(map.entries().to_vec())),
```

Comment:
We have an owned `MapRepresentation` here but we still clone. Maybe for the entries, we should provide an `MapRepresentation::into_entries` function that consumes the map so we can avoid cloning?